wreck incremental improvements

grondo opened the pull request:
This is a set of changes accumulated in wreck to support the splash app issue #1358. It still needs work and tests added, but features are functional, including:

 * Do not arbitrarily set nnodes in the kvs if the user didn't provide it, and preserve other user provided options in an `.opts` key in the kvs
 * Fix `flux-wreck purge` sort order, as well as removal of empty `lwj.x.y.z` dirs after a purge.
 * Add new `job.submit-nocreate` RPC to "submit" a job with an existing KVS entry, in support of:
 * Schedule `flux wreckrun` jobs if a scheduler is present and the new `-I, --immediate` option is not provided
 * Fix crash in `job.kvspath` #1394 
 * Distribute the `lwj.ntasks` global ntasks properly across all cores in wrexecd, instead of assuming `rank.X.cores` is the number of tasks to run on each rank `X`.
 * Add support for `R lite` format in `lwj.R_lite`. This  code expects `R_lite` to be an array of JSON objects with `rank:` keys (to be fixed soon if we need to switch to `HostName` keys)
 * `flux-wreckrun` now emits its manual "resource assignment" in R_lite format

Also thrown in (these should be pulled out probably):
 * resource-hwloc: Don't complain about "all ranks not aggregated" since the module may not be loaded everywhere
 * `flux-hwloc` Use FLUX_NODEID_ANY as target of get topo RPCs.

The re-worked code to parse `R_lite` falls back to `rank.X.cores` if there is no `R_lite` key, so it should work with the current scheduler and do a much better job of distributing tasks across the assigned resources, assuming that `rank.N.cores` is a count of assigned cores on each rank.

It may be that this approach which uses the current method of scheduler resource assignment will work "good enough" to satisfy #1358, but we'll need to have @trws comment. The only thing that a full `R_lite` implementation would gain is that we'd have the exact core list for each job, and wreck could set the default cpu binding for each job.

Lastly, I have to apologize for the quality of work here. This is definitely a rush job. Given more time it would have been better to refactor wrexecd.c first, and possibly the job module is due for a rewrite as well to remove all synchronous RPCs.

grondo commented:
Oops, #1400 reminded me that `flux wreck ls` breaks on this branch (at least the RANKS reporting does).

grondo commented:
This isn't ready for merge, something is wrong with the new `submit-only` RPC.

garlick commented on src/bindings/lua/wreck.lua:
```diff
@@ -315,10 +315,11 @@ end
 
```
Duplicate "the" in commit message (the the minimum of the number of...)

garlick commented on src/cmd/flux-wreckrun:
```diff
@@ -89,12 +90,16 @@ local function alloc_tasks (f, wreck, lwj)
     for i, ntasks in pairs (counts) do
-        local key = "rank."..i..".cores"
-        lwj [key] = ntasks
+        local corelist = "0"
```
Would it make sense here to keep the old rank.N.cores as an option for testing?

grondo replied:
we could I didn't want to add yet another option so in tests I just use --pre-launch-hook to do it. Easy to add an option though and that would be more straightforward.

garlick replied:
Oh as long as there is a way to drive it from tests.

grondo commented:
Ok, submit-only fixed and tested against flux-framework/flux-sched#295.

grondo commented:
Ok, fixed up the typo in commit messaged. I've also added @trws changes to add an `ncores` member to job requests, and published the `ncores` value in the submitted event message.

We'll need @dongahn to fix up his pending PR in flux-sched to also pull ncores directly from the event, and perhaps point me in the right direction to add the ncores field to jsc.

grondo commented:
> perhaps point me in the right direction to add the ncores field to jsc.

Ok, I'm going through jstatctl now adding NCORES everywhere there is also an NTASKS

grondo commented:
Ok, did my best to extend libjsc for `ncores` attribute of jobs. @dongahn should probably review.

dongahn commented:
> Ok, did my best to extend libjsc for ncores attribute of jobs. @dongahn should probably review.

Looks great! This shouldn't break the scheduler PR https://github.com/flux-framework/flux-sched/pull/295 as it still emits `ntasks`. So once this is merged, I will change sched code to use `scores` instead of `ntasks`.

garlick commented:
I like it!

grondo commented:
Thanks @garlick! Whew!
